modules/virt-creating-a-service-from-a-running-virtual-machine.adoc

```diff
@@ -0,0 +1,191 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/vm_networking/virt-using-the-default-pod-network-with-virt.adoc
+
+[id="virt-creating-a-service-from-a-running-virtual-machine_{context}"]
+
+= Creating a service from a running virtual machine
+
+You can create and expose a service from a running virtual machine.
+
+Once the virtual machine is running, create a service object to expose the service.
+Connect to the service by specifying the service name.
+
+The service exposes the host name of the virtual machine to the internal cluster,
+using the `ClusterIP` service, or externally, using the `NodePort` or `LoadBalancer`
+service.
+
+Because the service is scoped by the namespace to which it is defined, ensure
+that you reference the correct namespace in all YAML definitions.
+
+This procedure presents an example of how to create and expose a service object
+of `type: ClusterIP` as a virtual machine-backed service.
+
+[NOTE]
+====
+ClusterIP is the default service `type`, if the service `type` is not specified.
+====
+
+.Procedure
+
+. Start the virtual machine if it is not already running.
+
+. Edit the `VirtualMachine` YAML configuration file to include the
+label `special: key`.
++
+
+[NOTE]
+====
+Labels on a `VirtualMachine` are passed through to the pod. The labels on
+the `VirtualMachine`, for example `special: key`, must match the labels on
+the Service object `selector` attribute.
+====
+
++
+
+[source,yaml]
+----
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachine
+metadata:
+  name: vm-ephemeral
+  labels:
+    special: key <1>
+    spec:
+      domain:
+        devices:
+          disks:
+            - name: containerdisk
+              disk:
+                bus: virtio
+            - name: cloudinitdisk
+              disk:
+                bus: virtio
+          interfaces:
+          - masquerade: {}
+            name: default
+        resources:
+          requests:
+            memory: 1024M
+      networks:
+        - name: default
+          pod: {}
+      volumes:
+        - name: containerdisk
+          containerDisk:
+            image: kubevirt/fedora-cloud-container-disk-demo
+        - name: cloudinitdisk
+          cloudInitNoCloud:
+            userData: |
+              #!/bin/bash
+              echo "fedora" | passwd fedora --stdin
```

Comment:
This is still an invalid YAML. Applying it ends with an error:
```
$ oc apply -f vm-for-svc.yaml 
error: unable to decode "vm-for-svc.yaml": resource.metadataOnlyObject.ObjectMeta: v1.ObjectMeta.Labels: ReadString: expects " or n, but found {, error found in #10 byte of ...|:{"spec":{"domain":{|..., bigger context ...|ualMachineInstance","metadata":{"labels":{"spec":{"domain":{"devices":{"disks":[{"disk":{"bus":"virt|...
```

Reply:
I haven't verified it since my cluster is down, but I believe this should work:

```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: vm-ephemeral
spec:
  template:
    metadata:
      labels:
        special: key <1>
    spec:
      domain:
        devices:
          disks:
            - name: containerdisk
              disk:
                bus: virtio
            - name: cloudinitdisk
              disk:
                bus: virtio
          interfaces:
          - masquerade: {}
            name: default
        resources:
          requests:
            memory: 1024M
      networks:
        - name: default
          pod: {}
      volumes:
        - name: containerdisk
          containerDisk:
            image: kubevirt/fedora-cloud-container-disk-demo
        - name: cloudinitdisk
          cloudInitNoCloud:
            userData: |
              #!/bin/bash
              echo "fedora" | passwd fedora --stdin
```

Reply:
Thanks @phoracek --- I still have the question about whether we include a namespace attribute in this code in response to the comment from @maiqueb above. 

I added this statement:

_Because the service is scoped by the namespace to which it is defined, ensure
that you reference the correct namespace in all YAML definitions._

Is this statement true as I have worded it, and if so, should there not be a namespace attribute and what namespace should we supply?

@yossisegev this YAML has been updated with Petr's suggestion. Thanks.

Bob